File: sdk/client/src/stealth.ts
import { secp256k1 } from "@noble/curves/secp256k1";
import { keccak_256 } from "@noble/hashes/sha3";
import { type Hex, type Address, getAddress } from "viem";

// ── Types ──────────────────────────────────────────────────────────────────────

/** Compressed secp256k1 public key (33 bytes, hex-encoded with 0x prefix) */
export type CompressedPubKey = Hex;

/** Stealth meta-address: spending pubkey + viewing pubkey concatenated (66 bytes) */
export type StealthMetaAddress = Hex;

/** The "st:eth:0x..." URI format used in the stealth-meta auth-param */
export type StealthMetaURI = `st:eth:${Hex}`;

export interface StealthKeyPair {
  /** Private key (32 bytes) */
  privateKey: Hex;
  /** Compressed public key (33 bytes) */
  publicKey: CompressedPubKey;
}

export interface StealthKeys {
  spending: StealthKeyPair;
  viewing: StealthKeyPair;
}

export interface GenerateStealthAddressResult {
  /** The one-time stealth address to send funds to */
  stealthAddress: Address;
  /** Ephemeral public key to include in the Announcement (33 bytes compressed) */
  ephemeralPubKey: CompressedPubKey;
  /** View tag for fast scanning (single byte) */
  viewTag: number;
}

export interface StealthPaymentInfo {
  /** Detected stealth address that matches */
  stealthAddress: Address;
  /** Private key that controls the stealth address */
  stealthPrivateKey: Hex;
  /** Ephemeral public key from the announcement */
  ephemeralPubKey: CompressedPubKey;
}

// ── Helpers ────────────────────────────────────────────────────────────────────

/** Zero out a Uint8Array containing key material (H-TS-1). */
function zeroBytes(arr: Uint8Array): void {
  arr.fill(0);
}

function hexToBytes(hex: Hex): Uint8Array {
  const clean = hex.startsWith("0x") ? hex.slice(2) : hex;
  const bytes = new Uint8Array(clean.length / 2);
  for (let i = 0; i < bytes.length; i++) {
    bytes[i] = parseInt(clean.slice(i * 2, i * 2 + 2), 16);
  }
  return bytes;
}

function bytesToHex(bytes: Uint8Array): Hex {
  return `0x${Array.from(bytes)
    .map((b) => b.toString(16).padStart(2, "0"))
    .join("")}` as Hex;
}

function pubKeyToAddress(uncompressedPubKey: Uint8Array): Address {
  // Remove the 0x04 prefix (uncompressed point indicator)
  const pubKeyNoPrefix = uncompressedPubKey.slice(1);
  const hash = keccak_256(pubKeyNoPrefix);
  // Take last 20 bytes
  const addressBytes = hash.slice(12);
  return getAddress(bytesToHex(addressBytes));
}

// ── Core Functions ─────────────────────────────────────────────────────────────

/**
 * Generate a new stealth key pair (spending + viewing keys).
 * @returns The key pairs and the 66-byte stealth meta-address.
 */
export function generateStealthKeys(): {
  keys: StealthKeys;
  metaAddress: StealthMetaAddress;
} {
  const spendingPriv = secp256k1.utils.randomPrivateKey();
  const viewingPriv = secp256k1.utils.randomPrivateKey();

  const spendingPub = secp256k1.getPublicKey(spendingPriv, true); // compressed
  const viewingPub = secp256k1.getPublicKey(viewingPriv, true); // compressed

  const keys: StealthKeys = {
    spending: {
      privateKey: bytesToHex(spendingPriv),
      publicKey: bytesToHex(spendingPub),
    },
    viewing: {
      privateKey: bytesToHex(viewingPriv),
      publicKey: bytesToHex(viewingPub),
    },
  };

  // Meta-address = spendingPubKey (33) + viewingPubKey (33) = 66 bytes
  const metaBytes = new Uint8Array(66);
  metaBytes.set(spendingPub, 0);
  metaBytes.set(viewingPub, 33);

  const result = {
    keys,
    metaAddress: bytesToHex(metaBytes) as StealthMetaAddress,
  };

  // H-TS-1: Zero intermediate key material
  zeroBytes(spendingPriv);
  zeroBytes(viewingPriv);

  return result;
}

/**
 * Parse a stealth meta-address URI ("st:eth:0x...") into its component public keys.
 */
export function parseStealthMetaURI(uri: StealthMetaURI): {
  spendingPubKey: CompressedPubKey;
  viewingPubKey: CompressedPubKey;
} {
  if (!uri.startsWith("st:eth:0x")) {
    throw new Error(`Invalid stealth meta URI: ${uri}`);
  }
  const hex = uri.slice(7) as Hex; // Remove "st:eth:"
  return parseStealthMetaAddress(hex);
}

/**
 * Parse a raw stealth meta-address hex into its component public keys.
 */
export function parseStealthMetaAddress(metaAddress: StealthMetaAddress): {
  spendingPubKey: CompressedPubKey;
  viewingPubKey: CompressedPubKey;
} {
  const bytes = hexToBytes(metaAddress);
  if (bytes.length !== 66) {
    throw new Error(
      `Invalid stealth meta-address length: ${bytes.length}, expected 66`
    );
  }

  return {
    spendingPubKey: bytesToHex(bytes.slice(0, 33)),
    viewingPubKey: bytesToHex(bytes.slice(33, 66)),
  };
}

/**
 * Format a stealth meta-address as a URI for the stealth-meta auth-param.
 */
export function formatStealthMetaURI(
  metaAddress: StealthMetaAddress
): StealthMetaURI {
  return `st:eth:${metaAddress}` as StealthMetaURI;
}

/**
 * Derive a one-time stealth address from a recipient's stealth meta-address.
 * Called by the PAYER when making a confidential payment.
 *
 * Algorithm:
 * 1. Generate ephemeral key pair (r, R) where R = r * G
 * 2. Compute shared secret S = r * K_view
 * 3. viewTag = first byte of keccak256(S.x in compressed form)
 * 4. s = keccak256(S.x in compressed form) as scalar
 * 5. K_stealth = K_spend + s * G
 * 6. stealthAddr = address from K_stealth
 */
export function generateStealthAddress(
  spendingPubKey: CompressedPubKey,
  viewingPubKey: CompressedPubKey
): GenerateStealthAddressResult {
  // Generate ephemeral key pair
  const ephemeralPriv = secp256k1.utils.randomPrivateKey();
  const ephemeralPub = secp256k1.getPublicKey(ephemeralPriv, true); // compressed

  // Parse recipient's viewing public key
  const viewingPubBytes = hexToBytes(viewingPubKey);
  const viewingPoint = secp256k1.ProjectivePoint.fromHex(viewingPubBytes);

  // Compute shared secret: S = r * K_view
  const sharedPoint = viewingPoint.multiply(
    bytesToBigInt(ephemeralPriv)
  );
  const sharedCompressed = sharedPoint.toRawBytes(true); // compressed

  // Hash the shared secret
  const sharedHash = keccak_256(sharedCompressed);

  // View tag = first byte of hash
  const viewTag = sharedHash[0];

  // s = hash interpreted as scalar (mod n)
  const s = bytesToBigInt(sharedHash) % secp256k1.CURVE.n;

  // C-TS-2: Reject degenerate zero scalar (stealth address would equal spending pubkey)
  if (s === 0n) {
    throw new Error("Degenerate shared secret scalar (zero after reduction)");
  }

  // K_stealth = K_spend + s * G
  const spendingPubBytes = hexToBytes(spendingPubKey);
  const spendingPoint = secp256k1.ProjectivePoint.fromHex(spendingPubBytes);
  const sTimesG = secp256k1.ProjectivePoint.BASE.multiply(s);
  const stealthPoint = spendingPoint.add(sTimesG);

  // Compute address from uncompressed stealth public key
  const stealthPubUncompressed = stealthPoint.toRawBytes(false); // uncompressed (65 bytes)
  const stealthAddress = pubKeyToAddress(stealthPubUncompressed);

  const result = {
    stealthAddress,
    ephemeralPubKey: bytesToHex(ephemeralPub),
    viewTag,
  };

  // H-TS-1: Zero intermediate key material
  zeroBytes(ephemeralPriv);
  zeroBytes(sharedCompressed);
  zeroBytes(sharedHash);

  return result;
}

/**
 * Check if an announcement is addressed to us (fast path: view tag check first).
 * Called by the RECIPIENT's scanning service.
 *
 * Algorithm:
 * 1. Compute S' = k_view * R (where R = ephemeralPubKey from event)
 * 2. Compute viewTag' = first byte of keccak256(S')
 * 3. If viewTag' != event.viewTag, return null (fast filter)
 * 4. s' = keccak256(S') as scalar
 * 5. K_stealth' = K_spend + s' * G
 * 6. If address(K_stealth') == event.stealthAddress, match found
 * 7. stealthPrivKey = k_spend + s'
 */
export function checkStealthAnnouncement(
  announcement: {
    schemeId: bigint;
    stealthAddress: Address;
    ephemeralPubKey: CompressedPubKey;
    viewTag: number;
  },
  viewingPrivateKey: Hex,
  spendingPubKey: CompressedPubKey
): StealthPaymentInfo | null {
  // Only support scheme 1
  if (announcement.schemeId !== 1n) return null;

  const viewingPrivBytes = hexToBytes(viewingPrivateKey);
  const ephemeralPubBytes = hexToBytes(announcement.ephemeralPubKey);

  // Parse ephemeral public key
  const ephemeralPoint = secp256k1.ProjectivePoint.fromHex(ephemeralPubBytes);

  // C-TS-3: Validate viewing private key scalar range
  const viewingPrivScalar = bytesToBigInt(viewingPrivBytes);
  if (viewingPrivScalar === 0n || viewingPrivScalar >= secp256k1.CURVE.n) {
    zeroBytes(viewingPrivBytes);
    return null;
  }

  // Compute shared secret: S' = k_view * R
  const sharedPoint = ephemeralPoint.multiply(viewingPrivScalar);
  const sharedCompressed = sharedPoint.toRawBytes(true);

  // Hash the shared secret
  const sharedHash = keccak_256(sharedCompressed);

  // Fast filter: check view tag
  const computedViewTag = sharedHash[0];
  if (computedViewTag !== announcement.viewTag) {
    zeroBytes(viewingPrivBytes);
    zeroBytes(sharedCompressed);
    zeroBytes(sharedHash);
    return null;
  }

  // s' = hash as scalar (mod n)
  const s = bytesToBigInt(sharedHash) % secp256k1.CURVE.n;

  // C-TS-2: Reject degenerate zero scalar
  if (s === 0n) {
    return null;
  }

  // K_stealth' = K_spend + s' * G
  const spendingPubBytes = hexToBytes(spendingPubKey);
  const spendingPoint = secp256k1.ProjectivePoint.fromHex(spendingPubBytes);
  const sTimesG = secp256k1.ProjectivePoint.BASE.multiply(s);
  const stealthPoint = spendingPoint.add(sTimesG);

  // Compute address
  const stealthPubUncompressed = stealthPoint.toRawBytes(false);
  const computedAddress = pubKeyToAddress(stealthPubUncompressed);

  // Check if it matches the announced stealth address
  if (computedAddress.toLowerCase() !== announcement.stealthAddress.toLowerCase()) {
    return null;
  }

  // Compute stealth private key: k_stealth = k_spend + s' (mod n)
  // NOTE: We don't have spending private key here, so we compute it from
  // the viewing private key and return partial info. The full private key
  // computation needs computeStealthPrivateKey() with spending private key.
  // For now, we return a placeholder — the actual spending requires
  // the spending private key which the scanner may or may not have.
  
  // We compute the stealth private key component s for later use
  const sHex = `0x${s.toString(16).padStart(64, "0")}` as Hex;

  // H-TS-1: Zero intermediate key material
  zeroBytes(viewingPrivBytes);
  zeroBytes(sharedCompressed);
  zeroBytes(sharedHash);

  return {
    stealthAddress: computedAddress,
    stealthPrivateKey: sHex, // This is the shared secret scalar, needs k_spend added
    ephemeralPubKey: announcement.ephemeralPubKey,
  };
}

/**
 * Compute the full stealth private key (requires both viewing and spending private keys).
 * Used when the recipient wants to withdraw/spend from a stealth address.
 *
 * k_stealth = k_spend + keccak256(k_view * R) mod n
 */
export function computeStealthPrivateKey(
  spendingPrivateKey: Hex,
  ephemeralPubKey: CompressedPubKey,
  viewingPrivateKey: Hex
): Hex {
  const viewingPrivBytes = hexToBytes(viewingPrivateKey);
  const ephemeralPubBytes = hexToBytes(ephemeralPubKey);

  // Parse ephemeral public key
  const ephemeralPoint = secp256k1.ProjectivePoint.fromHex(ephemeralPubBytes);

  // C-TS-3: Validate key scalar ranges
  const viewingPrivScalar = bytesToBigInt(viewingPrivBytes);
  if (viewingPrivScalar === 0n || viewingPrivScalar >= secp256k1.CURVE.n) {
    throw new Error("Invalid viewing private key: out of scalar range");
  }

  // Compute shared secret: S = k_view * R
  const sharedPoint = ephemeralPoint.multiply(viewingPrivScalar);
  const sharedCompressed = sharedPoint.toRawBytes(true);

  // s = keccak256(S) mod n
  const sharedHash = keccak_256(sharedCompressed);
  const s = bytesToBigInt(sharedHash) % secp256k1.CURVE.n;

  // C-TS-2: Reject degenerate zero scalar
  if (s === 0n) {
    throw new Error("Degenerate shared secret scalar (zero after reduction)");
  }

  // k_stealth = k_spend + s mod n
  const spendingPrivBytes = hexToBytes(spendingPrivateKey);
  const kSpend = bytesToBigInt(spendingPrivBytes);
  if (kSpend === 0n || kSpend >= secp256k1.CURVE.n) {
    throw new Error("Invalid spending private key: out of scalar range");
  }
  const kStealth = (kSpend + s) % secp256k1.CURVE.n;

  const result = `0x${kStealth.toString(16).padStart(64, "0")}` as Hex;

  // H-TS-1: Zero intermediate key material
  zeroBytes(viewingPrivBytes);
  zeroBytes(spendingPrivBytes);
  zeroBytes(sharedCompressed);
  zeroBytes(sharedHash);

  return result;
}

// ── Utility ────────────────────────────────────────────────────────────────────

function bytesToBigInt(bytes: Uint8Array): bigint {
  let result = 0n;
  for (const byte of bytes) {
    result = (result << 8n) | BigInt(byte);
  }
  return result;
}

File: sdk/python/pympp/stealth.py
"""Core stealth address cryptography for ERC-5564 scheme ID 1 (secp256k1).

Uses coincurve (libsecp256k1 wrapper) for constant-time EC operations.
"""

from __future__ import annotations

import os
import hmac
from dataclasses import dataclass
from typing import Optional

import coincurve
from Crypto.Hash import keccak as _keccak

# secp256k1 curve order
_N = 0xFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFEBAAEDCE6AF48A03BBFD25E8CD0364141


def _keccak256(data: bytes) -> bytes:
    h = _keccak.new(digest_bits=256)
    h.update(data)
    return h.digest()


def _bytes_to_int(b: bytes) -> int:
    return int.from_bytes(b, "big")


def _int_to_bytes32(n: int) -> bytes:
    return n.to_bytes(32, "big")


def _hex_to_bytes(hex_str: str) -> bytes:
    clean = hex_str[2:] if hex_str.startswith("0x") else hex_str
    return bytes.fromhex(clean)


def _bytes_to_hex(b: bytes) -> str:
    return "0x" + b.hex()


def _pub_key_to_address(uncompressed_pub: bytes) -> str:
    """Derive checksummed Ethereum address from uncompressed public key (65 bytes)."""
    pub_no_prefix = uncompressed_pub[1:]
    h = _keccak256(pub_no_prefix)
    addr_bytes = h[12:]
    return _checksum_address("0x" + addr_bytes.hex())


def _checksum_address(address: str) -> str:
    """EIP-55 checksum encoding."""
    addr = address.lower().replace("0x", "")
    hash_hex = _keccak256(addr.encode("ascii")).hex()
    result = "0x"
    for i, c in enumerate(addr):
        if c in "0123456789":
            result += c
        elif int(hash_hex[i], 16) >= 8:
            result += c.upper()
        else:
            result += c.lower()
    return result


def _constant_time_compare(a: str, b: str) -> bool:
    """Constant-time string comparison for addresses."""
    return hmac.compare_digest(a.lower(), b.lower())


def _generate_valid_private_key() -> bytes:
    """Generate a random 32-byte private key that is valid for secp256k1."""
    while True:
        key_bytes = os.urandom(32)
        k = _bytes_to_int(key_bytes)
        if 0 < k < _N:
            return key_bytes


def _scalar_mult(point_bytes: bytes, scalar_bytes: bytes) -> coincurve.PublicKey:
    """Multiply an EC point by a scalar using coincurve (constant-time)."""
    pub = coincurve.PublicKey(point_bytes)
    result = pub.multiply(scalar_bytes)
    return result


# ── Types ──────────────────────────────────────────────────────────────────────


@dataclass
class StealthKeyPair:
    private_key: str  # 0x-prefixed 32-byte hex
    public_key: str  # 0x-prefixed 33-byte compressed hex


@dataclass
class StealthKeys:
    spending: StealthKeyPair
    viewing: StealthKeyPair


@dataclass
class GenerateStealthAddressResult:
    stealth_address: str
    ephemeral_pub_key: str  # compressed 33-byte hex
    view_tag: int  # 0-255


@dataclass
class StealthPaymentInfo:
    stealth_address: str
    stealth_private_key: str  # shared secret scalar (needs k_spend added)
    ephemeral_pub_key: str


# ── Core Functions ─────────────────────────────────────────────────────────────


def generate_stealth_keys() -> tuple[StealthKeys, str]:
    """Generate stealth key pairs and 66-byte meta-address.

    Returns:
        Tuple of (StealthKeys, meta_address_hex).
    """
    spending_priv = _generate_valid_private_key()
    viewing_priv = _generate_valid_private_key()

    spending_pub = coincurve.PublicKey.from_secret(spending_priv).format(compressed=True)
    viewing_pub = coincurve.PublicKey.from_secret(viewing_priv).format(compressed=True)

    keys = StealthKeys(
        spending=StealthKeyPair(
            private_key=_bytes_to_hex(spending_priv),
            public_key=_bytes_to_hex(spending_pub),
        ),
        viewing=StealthKeyPair(
            private_key=_bytes_to_hex(viewing_priv),
            public_key=_bytes_to_hex(viewing_pub),
        ),
    )

    meta_bytes = spending_pub + viewing_pub  # 66 bytes
    meta_address = _bytes_to_hex(meta_bytes)

    return keys, meta_address


def parse_stealth_meta_address(meta_address: str) -> tuple[str, str]:
    """Parse a 66-byte stealth meta-address into spending and viewing public keys.

    Returns:
        Tuple of (spending_pub_key_hex, viewing_pub_key_hex).
    """
    b = _hex_to_bytes(meta_address)
    if len(b) != 66:
        raise ValueError(f"Invalid stealth meta-address length: {len(b)}, expected 66")

    spending_pub = _bytes_to_hex(b[:33])
    viewing_pub = _bytes_to_hex(b[33:66])
    return spending_pub, viewing_pub


def parse_stealth_meta_uri(uri: str) -> tuple[str, str]:
    """Parse 'st:eth:0x...' URI into (spending_pub_key, viewing_pub_key)."""
    if not uri.startswith("st:eth:0x"):
        raise ValueError(f"Invalid stealth meta URI: {uri}")
    hex_part = uri[7:]  # Remove "st:eth:"
    return parse_stealth_meta_address(hex_part)


def format_stealth_meta_uri(meta_address: str) -> str:
    """Format a meta-address as 'st:eth:0x...' URI."""
    return f"st:eth:{meta_address}"


def generate_stealth_address(
    spending_pub_key: str,
    viewing_pub_key: str,
) -> GenerateStealthAddressResult:
    """Derive a one-time stealth address from recipient's public keys.

    Called by the PAYER.

    Algorithm:
        1. Generate ephemeral key pair (r, R = r*G)
        2. S = r * K_view (shared secret)
        3. H = keccak256(compress(S))
        4. viewTag = H[0]
        5. s = H mod n
        6. K_stealth = K_spend + s*G
        7. stealthAddr = address(K_stealth)
    """
    # Generate ephemeral key pair
    ephemeral_priv_bytes = _generate_valid_private_key()
    ephemeral_pub = coincurve.PublicKey.from_secret(ephemeral_priv_bytes).format(compressed=True)

    # Parse viewing public key
    viewing_pub_bytes = _hex_to_bytes(viewing_pub_key)

    # Shared secret: S = r * K_view (ECDH)
    shared_pub = _scalar_mult(viewing_pub_bytes, ephemeral_priv_bytes)
    shared_compressed = shared_pub.format(compressed=True)

    # Hash the shared secret
    shared_hash = _keccak256(shared_compressed)

    # View tag = first byte
    view_tag = shared_hash[0]

    # s = hash as scalar mod n
    s = _bytes_to_int(shared_hash) % _N

    # Reject degenerate zero scalar
    if s == 0:
        raise ValueError("Degenerate shared secret scalar (zero after reduction)")

    # K_stealth = K_spend + s*G
    spending_pub_bytes = _hex_to_bytes(spending_pub_key)
    spending_pub = coincurve.PublicKey(spending_pub_bytes)
    s_times_g = coincurve.PublicKey.from_secret(_int_to_bytes32(s))
    stealth_pub = coincurve.PublicKey.combine_keys([spending_pub, s_times_g])

    # Address from uncompressed stealth public key
    stealth_pub_uncompressed = stealth_pub.format(compressed=False)
    stealth_address = _pub_key_to_address(stealth_pub_uncompressed)

    return GenerateStealthAddressResult(
        stealth_address=stealth_address,
        ephemeral_pub_key=_bytes_to_hex(ephemeral_pub),
        view_tag=view_tag,
    )


def check_stealth_announcement(
    announcement: dict,
    viewing_private_key: str,
    spending_pub_key: str,
) -> Optional[StealthPaymentInfo]:
    """Check if an announcement is addressed to us.

    Args:
        announcement: Dict with schemeId, stealthAddress, ephemeralPubKey, viewTag.
        viewing_private_key: Hex viewing private key.
        spending_pub_key: Hex compressed spending public key.

    Returns:
        StealthPaymentInfo if match, None otherwise.
    """
    if announcement["schemeId"] != 1:
        return None

    viewing_priv_bytes = _hex_to_bytes(viewing_private_key)
    viewing_scalar = _bytes_to_int(viewing_priv_bytes)

    # Validate viewing key scalar range
    if viewing_scalar == 0 or viewing_scalar >= _N:
        return None

    ephemeral_pub_bytes = _hex_to_bytes(announcement["ephemeralPubKey"])

    # S' = k_view * R (constant-time via coincurve/libsecp256k1)
    try:
        shared_pub = _scalar_mult(ephemeral_pub_bytes, viewing_priv_bytes)
    except Exception:
        return None
    shared_compressed = shared_pub.format(compressed=True)

    shared_hash = _keccak256(shared_compressed)

    # Fast filter: check view tag
    computed_view_tag = shared_hash[0]
    if computed_view_tag != announcement["viewTag"]:
        return None

    # s' = hash mod n
    s = _bytes_to_int(shared_hash) % _N

    # Reject degenerate zero scalar
    if s == 0:
        return None

    # K_stealth' = K_spend + s'*G
    spending_pub_bytes = _hex_to_bytes(spending_pub_key)
    spending_pub = coincurve.PublicKey(spending_pub_bytes)
    s_times_g = coincurve.PublicKey.from_secret(_int_to_bytes32(s))
    stealth_pub = coincurve.PublicKey.combine_keys([spending_pub, s_times_g])

    stealth_pub_uncompressed = stealth_pub.format(compressed=False)
    computed_address = _pub_key_to_address(stealth_pub_uncompressed)

    # Constant-time address comparison
    if not _constant_time_compare(computed_address, announcement["stealthAddress"]):
        return None

    s_hex = "0x" + format(s, "064x")

    return StealthPaymentInfo(
        stealth_address=computed_address,
        stealth_private_key=s_hex,
        ephemeral_pub_key=announcement["ephemeralPubKey"],
    )


def compute_stealth_private_key(
    spending_private_key: str,
    ephemeral_pub_key: str,
    viewing_private_key: str,
) -> str:
    """Compute full stealth private key for spending.

    k_stealth = k_spend + keccak256(k_view * R) mod n
    """
    viewing_priv_bytes = _hex_to_bytes(viewing_private_key)
    viewing_scalar = _bytes_to_int(viewing_priv_bytes)

    # Validate key scalar ranges
    if viewing_scalar == 0 or viewing_scalar >= _N:
        raise ValueError("Invalid viewing private key: out of scalar range")

    ephemeral_pub_bytes = _hex_to_bytes(ephemeral_pub_key)

    # S = k_view * R (constant-time)
    shared_pub = _scalar_mult(ephemeral_pub_bytes, viewing_priv_bytes)
    shared_compressed = shared_pub.format(compressed=True)

    # s = keccak256(S) mod n
    shared_hash = _keccak256(shared_compressed)
    s = _bytes_to_int(shared_hash) % _N

    if s == 0:
        raise ValueError("Degenerate shared secret scalar (zero after reduction)")

    # k_stealth = k_spend + s mod n
    spending_priv_bytes = _hex_to_bytes(spending_private_key)
    k_spend = _bytes_to_int(spending_priv_bytes)

    if k_spend == 0 or k_spend >= _N:
        raise ValueError("Invalid spending private key: out of scalar range")

    k_stealth = (k_spend + s) % _N

    return "0x" + format(k_stealth, "064x")
